utils/SolanaAuth.ts

```diff
@@ -0,0 +1,127 @@
+import base58 from "bs58";
+import util from "tweetnacl-util";
+import { sign } from "tweetnacl";
+import { signInMessage } from "../config";
+
+export interface Adapter {
+  updateNonce(pubkey: string): Promise<string>;
+  createProfile(pubkey: string): Promise<string>;
+  checkNonce(pubkey: string): Promise<any>;
+  generateToken(pubkey: string): Promise<string>;
+  verifyTTL(pubkey: string): Promise<boolean>;
+}
+
+interface SolanaAuth {
+  getSolanaAuth: (req: any, res: any) => Promise<void>;
+  completeSolanaAuth: (req: any, res: any) => Promise<void>;
+}
+type SolanaAuthOptions = {
+  adapter: Adapter;
+};
+
+// TODO: Turn this into a function that can be used as the defualt export on a Next.JS Auth page.
+const SolanaAuth = (options: SolanaAuthOptions): SolanaAuth => {
+  const getNonce = async (pubkey: string) => {
+    /**
+     * If there is NO nonce, create the user profile with a new nonce
+     * if there is a nonce, update it.
+     */
+
+    let nonce = await options.adapter.checkNonce(pubkey);
+
+    if (!nonce) {
+      nonce = await options.adapter.createProfile(pubkey);
+    } else {
+      nonce = await options.adapter.updateNonce(pubkey);
+    }
+    return nonce;
+  };
+
+  const getSolanaAuth = async (req: any, res: any) => {
+    // get pubkey form req
+    const pubkey = req.query.pubkey;
+    if (pubkey) {
+      let nonce = await getNonce(pubkey.toString());
+
+      res.status(200).json({ nonce });
+    } else {
+      res.status(400).json("No public key specified");
+    }
+  };
+
+  const completeSolanaAuth = async (req: any, res: any) => {
+    try {
+      // parse the query parameters
+      let { pubkey, payload, signature } = req.query;
+      payload = payload.toString();
+      pubkey = pubkey.toString();
+
+      // verify the TLL
+      const ttlVerified = options.adapter.verifyTTL(pubkey);
+      if (!ttlVerified) throw new Error("Nonce is expired");
+
+      // get the nonce from the database
+      let dbNonce = await options.adapter.checkNonce(pubkey);
+      if (!dbNonce) throw new Error("Public Key not in DB");
+
+      const { nonce, domain } = parsePayload(payload);
+
+      // verify the payload
+      const constructedMessage = signInMessage(nonce, domain);
+
+      if (constructedMessage !== payload) throw new Error("Invalid payload");
+
+      // TODO: verify the domain (dynamically)
+      // check nonce against nonce in db
+      if (nonce !== dbNonce) throw new Error("Nonce is invalid");
+
+      payload = util.decodeUTF8(payload);
+      const publicKey = base58.decode(pubkey);
+      signature = qptua(signature);
+
+      // verify that the bytes were signed witht the private key
+      if (!sign.detached.verify(payload, signature, publicKey))
+        throw new Error("invalid signature");
+
+      // TODO: Create the JWT token with Firebase and send to client
```

Comment:
You brought back this TODO and others from the dead 🧟‍♀️

Mind fixing and double checking you didn't bring in any other old code?